llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -3487,6 +3487,67 @@ Instruction *InstCombinerImpl::foldICmpBinOpEqualityWithConstant(
       Value *And = Builder.CreateAnd(BOp0, NotBOC);
       return new ICmpInst(Pred, And, NotBOC);
     }
+    // (icmp eq (or (select cond, 0, NonZero), Other))
+    //  -> (and cond, (icmp eq Other, 0))
+    // (icmp ne (or (select cond, NonZero, 0), Other))
+    //  -> (or cond, (icmp ne Other, 0))
+    // (icmp ne (or (select cond, 0, NonZero), Other))
+    //  -> (or (not cond), (icmp ne Other, 0))
+    // (icmp eq (or (select cond, NonZero, 0), Other))
+    //  -> (and (not cond), (icmp eq Other, 0))
+    Value *Cond, *TV, *FV, *Other;
+    if (C.isZero() &&
+        match(BO, m_c_Or(m_Select(m_Value(Cond), m_Value(TV), m_Value(FV)),
+                         m_Value(Other)))) {
+      const SimplifyQuery Q = SQ.getWithInstruction(&Cmp);
+      auto IsNonZero = [&](Value *V) {
+        return isKnownNonZero(V, Q.DL, /*Depth=*/0, Q.AC, Q.CxtI, Q.DT);
+      };
+      // Easy case is if eq/ne matches whether 0 is trueval/falseval.
+      if (Pred == ICmpInst::ICMP_EQ
+              ? (match(TV, m_SpecificInt(C)) && IsNonZero(FV))
+              : (match(FV, m_SpecificInt(C)) && IsNonZero(TV))) {
+        Value *Cmp = Builder.CreateICmp(
+            Pred, Other, Constant::getNullValue(Other->getType()));
+        return BinaryOperator::Create(
+            Pred == ICmpInst::ICMP_EQ ? Instruction::And : Instruction::Or, Cmp,
+            Cond);
+      }
+      // Harder case is if eq/ne matches whether 0 is falseval/trueval. In this
+      // case we need to invert the select condition so we need to be careful to
+      // avoid creating extra instructions.
```

Comment:
Is all this complexity needed for motivating patterns? Can we only handle one-use BO + select instead?

Reply:
I'm happy to replace the `canFreelyInvertAllUses`/`freelyInvertAlluses` with just `getFreelyInverted`. @dtcxzyw does that cover the case you saw?

Reply:
FWIW we will only be saving 5/6 LOC.

Reply:
The freelyInvertAllUses() is the main thing I'd want to get rid of, yes.

Reply:
@dtcxzyw does the case you had req inverting all uses or is that unnecessary?

Reply:
I prefer to create extra not instructions and add one-use constraints on BO and select.

All occurrences of this pattern in my benchmark:
```
; 77 occurrences:
; fmt/optimized/format.cc.ll
; fmt/optimized/gtest-extra.cc.ll
; fmt/optimized/header-only-test.cc.ll
; fmt/optimized/os.cc.ll
; fmt/optimized/unicode-test.cc.ll
; fmt/optimized/util.cc.ll
; lief/optimized/ASN1Reader.cpp.ll
; lief/optimized/AndroidIdent.cpp.ll
; lief/optimized/BinaryParser.cpp.ll
; lief/optimized/BuildToolVersion.cpp.ll
; lief/optimized/Builder.cpp.ll
; lief/optimized/ChainedFixup.cpp.ll
; lief/optimized/CodeIntegrity.cpp.ll
; lief/optimized/CodeView.cpp.ll
; lief/optimized/CodeViewPDB.cpp.ll
; lief/optimized/CoreAuxv.cpp.ll
; lief/optimized/CoreFile.cpp.ll
; lief/optimized/CorePrPsInfo.cpp.ll
; lief/optimized/CorePrStatus.cpp.ll
; lief/optimized/CoreSigInfo.cpp.ll
; lief/optimized/Debug.cpp.ll
; lief/optimized/DelayImport.cpp.ll
; lief/optimized/DelayImportEntry.cpp.ll
; lief/optimized/DyldChainedFixups.cpp.ll
; lief/optimized/DyldInfo.cpp.ll
; lief/optimized/DynamicEntryArray.cpp.ll
; lief/optimized/DynamicEntryLibrary.cpp.ll
; lief/optimized/DynamicEntryRpath.cpp.ll
; lief/optimized/DynamicEntryRunPath.cpp.ll
; lief/optimized/DynamicSharedObject.cpp.ll
; lief/optimized/File.cpp.ll
; lief/optimized/FileStream.cpp.ll
; lief/optimized/GenericContent.cpp.ll
; lief/optimized/LangCodeItem.cpp.ll
; lief/optimized/LinkEdit.cpp.ll
; lief/optimized/LoadConfiguration.cpp.ll
; lief/optimized/LoadConfigurationV0.cpp.ll
; lief/optimized/LoadConfigurationV1.cpp.ll
; lief/optimized/LoadConfigurationV10.cpp.ll
; lief/optimized/LoadConfigurationV11.cpp.ll
; lief/optimized/LoadConfigurationV3.cpp.ll
; lief/optimized/LoadConfigurationV4.cpp.ll
; lief/optimized/LoadConfigurationV5.cpp.ll
; lief/optimized/LoadConfigurationV6.cpp.ll
; lief/optimized/LoadConfigurationV7.cpp.ll
; lief/optimized/LoadConfigurationV8.cpp.ll
; lief/optimized/LoadConfigurationV9.cpp.ll
; lief/optimized/MemoryStream.cpp.ll
; lief/optimized/Note.cpp.ll
; lief/optimized/NoteAbi.cpp.ll
; lief/optimized/NoteGnuProperty.cpp.ll
; lief/optimized/PKCS9SigningTime.cpp.ll
; lief/optimized/Pogo.cpp.ll
; lief/optimized/PogoEntry.cpp.ll
; lief/optimized/RelocationEntry.cpp.ll
; lief/optimized/RelocationFixup.cpp.ll
; lief/optimized/Repro.cpp.ll
; lief/optimized/ResourceNode.cpp.ll
; lief/optimized/ResourceStringFileInfo.cpp.ll
; lief/optimized/ResourceVarFileInfo.cpp.ll
; lief/optimized/ResourceVersion.cpp.ll
; lief/optimized/ResourcesManager.cpp.ll
; lief/optimized/ResourcesParser.cpp.ll
; lief/optimized/RichHeader.cpp.ll
; lief/optimized/Signature.cpp.ll
; lief/optimized/SignatureParser.cpp.ll
; lief/optimized/SignerInfo.cpp.ll
; lief/optimized/SpanStream.cpp.ll
; lief/optimized/SpcIndirectData.cpp.ll
; lief/optimized/StackSize.cpp.ll
; lief/optimized/TLS.cpp.ll
; lief/optimized/TrieNode.cpp.ll
; lief/optimized/Type.cpp.ll
; lief/optimized/VectorStream.cpp.ll
; lief/optimized/hash_stream.cpp.ll
; lief/optimized/x509.cpp.ll
; spdlog/optimized/bundled_fmtlib_format.cpp.ll
define i1 @func000000000000121b(i1 %0, i32 %1, i1 %2) {
entry:
  %3 = icmp sgt i32 %1, -1
  call void @llvm.assume(i1 %3)
  %4 = zext i1 %2 to i32
  %5 = select i1 %0, i32 %1, i32 0
  %6 = or i32 %5, %4
  %7 = icmp eq i32 %6, 0
  ret i1 %7
}

; 1 occurrences:
; libquic/optimized/time_posix.cc.ll
define i1 @func0000000501007f1b(i8 %0, i64 %1, i1 %2) {
entry:
  %3 = select i1 %2, i8 2, i8 1
  %4 = icmp sgt i64 %1, -1
  %5 = select i1 %4, i8 0, i8 %3
  %6 = or i8 %5, %0
  %7 = icmp eq i8 %6, 0
  ret i1 %7
}

; 1 occurrences:
; linux/optimized/intel_hdcp.ll
define i1 @func000000000000411b(i32 %0, i1 %1) {
entry:
  %2 = select i1 %1, i32 0, i32 -110
  %3 = or i32 %2, %0
  %4 = icmp eq i32 %3, 0
  ret i1 %4
}

; 1 occurrences:
; qemu/optimized/tcg.c.ll
define i1 @func00000000051067fb(i1 %0, i32 %1) {
entry:
  %2 = lshr i32 %1, 3
  %3 = and i32 %2, 1
  call void @use.t0(i32 %3)
  %4 = select i1 %0, i32 16384, i32 0
  %5 = or disjoint i32 %3, %4
  %6 = icmp eq i32 %5, 0
  ret i1 %6
}

declare void @use.t0(i32)

; 1 occurrences:
; qemu/optimized/hw_net_cadence_gem.c.ll
define i1 @func00000000122c08db(i64 %0, i64 %1, i32 %2) {
entry:
  %3 = icmp sgt i64 %0, -1
  call void @llvm.assume(i1 %3)
  %4 = sub i32 0, %2
  %5 = and i32 %2, %4
  %6 = icmp sgt i32 %5, -1
  call void @llvm.assume(i1 %6)
  %7 = icmp eq i32 %5, 0
  %8 = select i1 %7, i64 0, i64 %1
  %9 = or disjoint i64 %8, %0
  %10 = icmp eq i64 %9, 0
  ret i1 %10
}

; 1 occurrences:
; qemu/optimized/util_host-utils.c.ll
define i1 @func000000000000841b(i64 %0, i1 noundef %1, i64 %2) {
entry:
  %3 = select i1 %1, i64 0, i64 %2
  %4 = or i64 %3, %0
  %5 = icmp eq i64 %4, 0
  ret i1 %5
}
```

I will make the dataset public this month :)

Reply:
Okay, removed all the `canfreelyinvertalluses` logic. I keep a check for `getFreelyInverted` if the sel is not single-use as thats pretty trivial.